vertical-pod-autoscaler/pkg/utils/metrics/recommender/recommender_test.go

```diff
@@ -0,0 +1,310 @@
+				{
+					Recommendation: &vpa_types.RecommendedPodResources{
+						ContainerRecommendations: nil,
+					},
+				},
+			},
+			wantMetrics: map[string]float64{
+				"api=v1,has_recommendation=false,matches_pods=true,unsupported_config=false,update_mode=Auto,": 1,
+			},
+		},
+		{
+			name: "report has recommendation as true on existing container recommendations",
+			add: []*model.Vpa{
+				{
+					Recommendation: &vpa_types.RecommendedPodResources{
+						ContainerRecommendations: []vpa_types.RecommendedContainerResources{{}},
+					},
+				},
+			},
+			wantMetrics: map[string]float64{
+				"api=v1,has_recommendation=true,matches_pods=true,unsupported_config=false,update_mode=Auto,": 1,
+			},
+		},
+		{
+			name: "report has matches pods as true on missing condition",
+			add: []*model.Vpa{
+				{
+					Conditions: nil,
+				},
+			},
+			wantMetrics: map[string]float64{
+				"api=v1,has_recommendation=false,matches_pods=true,unsupported_config=false,update_mode=Auto,": 1,
+			},
+		},
+		{
+			name: "report has matches pods as false on NoPodsMatched condition",
+			add: []*model.Vpa{
+				{
+					Conditions: map[vpa_types.VerticalPodAutoscalerConditionType]vpa_types.VerticalPodAutoscalerCondition{
+						vpa_types.NoPodsMatched: {
+							Status: apiv1.ConditionTrue,
+						},
+					},
+				},
+			},
+			wantMetrics: map[string]float64{
+				"api=v1,has_recommendation=false,matches_pods=false,unsupported_config=false,update_mode=Auto,": 1,
+			},
+		},
+		{
+			name: "report unsupported config as false on missing condition",
+			add: []*model.Vpa{
+				{
+					Conditions: nil,
+				},
+			},
+			wantMetrics: map[string]float64{
+				"api=v1,has_recommendation=false,matches_pods=true,unsupported_config=false,update_mode=Auto,": 1,
+			},
+		},
+		{
+			name: "report unsupported config as false on ConfigUnsupported condition",
+			add: []*model.Vpa{
+				{
+					Conditions: map[vpa_types.VerticalPodAutoscalerConditionType]vpa_types.VerticalPodAutoscalerCondition{
+						vpa_types.ConfigUnsupported: {
+							Status: apiv1.ConditionTrue,
+						},
+					},
+				},
+			},
+			wantMetrics: map[string]float64{
+				"api=v1,has_recommendation=false,matches_pods=true,unsupported_config=true,update_mode=Auto,": 1,
+			},
+		},
+	}
+
+	for _, tc := range cases {
+		t.Run(tc.name, func(t *testing.T) {
+			counter := NewObjectCounter()
+			for _, add := range tc.add {
+				counter.Add(add)
+			}
+
+			t.Cleanup(func() {
+				// Reset the metric after the test to avoid collisions.
+				vpaObjectCount.Reset()
+			})
+			counter.Observe()
+
+			metrics := make(chan prometheus.Metric)
+
+			go func() {
+				vpaObjectCount.Collect(metrics)
+				close(metrics)
+			}()
+
+			gotMetrics := make(map[string]float64)
+			for metric := range metrics {
+				var metricProto dto.Metric
+				if err := metric.Write(&metricProto); err != nil {
+					t.Errorf("failed to write metric: %v", err)
+				}
+
+				key := labelsToKey(metricProto.GetLabel())
```

Comment:
I think it's better to compare `LabelPair`s directly rather than convert them to custom string representations

Reply:
I agree. Do you have a proposal how to record the metrics and define the cases in a still readable way?
Using cmp we could probably store the gotMetrics and wantMetrics as lists instead, but creating the test cases will be much harder to read.